pycloudlib/ec2/instance.py

```diff
@@ -81,9 +105,67 @@ def add_network_interface(self) -> str:
         https://boto3.readthedocs.io/en/latest/reference/services/ec2.html?#EC2.Client.attach_network_interface
+
+        Args:
+            ipv4_address_count: number of private ipv4s
+            ipv6_address_count: number of private ipv6s
+            ipv4_public_ip_count: number of public ips associated to private
+                ips
+            associate_public_ip: If True, associate a per-interface public
+                ipv4
+        Returns: private primary ip
         """
         self._log.debug("adding network interface to %s", self.id)
-        interface_id = self._create_network_interface()
+        interface_id = self._create_network_interface(ipv6_address_count)
+        if ipv4_address_count < ipv4_public_ip_count:
+            raise PycloudlibError(
+                f"Invalid configuration:"
+                f" `ipv4_public_ip_count={ipv4_public_ip_count}` cannot be"
+                f" greater than `ipv4_address_count={ipv4_address_count}`"
+            )
+
+        if ipv4_address_count is not None and ipv4_address_count > 1:
+            assignment_resp = self._client.assign_private_ip_addresses(
+                NetworkInterfaceId=interface_id,
+                # Minus one to account for the private_ipv4 that comes by
+                # default with the NIC
+                SecondaryPrivateIpAddressCount=ipv4_address_count - 1,
+            )
+            _check_response(assignment_resp)
+            if ipv4_public_ip_count > 1:
+                nics_resp = self._client.describe_network_interfaces(
+                    NetworkInterfaceIds=[interface_id]
+                )
+                _check_response(nics_resp)
+                priv_ips = [
+                    nics_resp["NetworkInterfaces"][0]["PrivateIpAddress"]
+                ]
+                priv_ips.extend(
+                    (
+                        x["PrivateIpAddress"]
+                        for x in assignment_resp["AssignedPrivateIpAddresses"]
+                    )
+                )
+                # Skip default NIC, as it already comes with a public ipv4
+                for i in range(1, ipv4_public_ip_count):
+                    allocation = self._client.allocate_address(Domain="vpc")
+                    _check_response(allocation)
+                    association = self._client.associate_address(
+                        AllocationId=allocation["AllocationId"],
+                        NetworkInterfaceId=interface_id,
+                        PrivateIpAddress=priv_ips[i],
+                    )
+                    _check_response(association)
+
+        if associate_public_ip:
+            allocation = self._client.allocate_address(Domain="vpc")
+            _check_response(allocation)
+            association = self._client.associate_address(
+                AllocationId=allocation["AllocationId"],
+                NetworkInterfaceId=interface_id,
+            )
+            _check_response(association)
```

Comment:
While this feels strange to provide a separate associate_public_ip, the use-case and the calls to the botocore API are different given we are trying to associate a public IP address to the primary private_IP on this interface_id. This works for the use-cases we have today and provides us with a simple mechanism to assign a single public IP to a given interface without having to also having to creation secondary private IPs.

Reply:
Thanks for the simplification.